text/3720-labeled-match.md

```diff
@@ -0,0 +1,1090 @@
+
+In theory, more sophisticated analysis of the MIR should be able to optimize the "loop + match" pattern into a collection of unconditional jumps. We've seen that it's not capable of performing this optimization today, but if it could, then from a performance perspective maybe labeled match would not be needed.
+
+While improvements to rust's MIR passes (or even a whole new IR that is better suited to optimization) are certainly possible, limitations are:
+
+- the implementation complexity
+- the compile time cost
+- analysis is fragile
+- this transform may not be adventageous in general
+
+In contrast
+
+- labeled match is a desugaring no more complex than labeled loops and blocks
+- the transformation is syntactic, and therefore nicely bounded
+- programmers can write their code in such a way that they can be confident the desugaring to a `goto` kicks in
+- the (expert) programmer definitely wants the desugaring into a `goto`
+
+So, labeled match is a solid way to make progress on better codegen. Improved optimizations on MIR are also very welcome, but never entirely remove the need for labeled match from a programmer's perspective.
+
+## recognize "loop + match" and optimize
+
+In theory it is possible to internally recognize and rewrite a "loop + match" expression into a labeled match. With this approach, no changes to language syntax are needed.
+
+A fundamental problem with this approach is a change in drop order:
+
+```rust
+let mut state = 0;
+'label: loop {
+    match state {
+        0 => {
+            let x = vec![1,2,3];
+            state = 1;
+            // drop of `x` gets inserted between state update and jump
+            continue 'label;
+        }
+        _ => ...
+    }
+}
+
+// versus if you rewrite to labeled match
+
+'label: match 0 {
+    0 => {
+        let x = vec![1,2,3];
+        // drop of `x` happens before the state update
+        continue 'label 1;
+    }
+    _ => ...
+}
+```
+
+Beyond that, the analysis for recognizing "loop + match" will likely be complex and fragile. Part of the appeal of labeled match is that the desugaring rules are simple and deterministic. Using a labeled match signals that something subtle is going on: for readers and future reviewers it is clear that the labeled match desugaring is desired and potentially crucial for the code to perform well.
+
+## Why Labeled match is the best solution
+
+Finally, let's summarize labeled match.
+
+The labeled match proposal combines existing rust features of `match` and labeled blocks/loops. It is just the interaction between these concepts that has to be learned, no new keywords or syntactic constructions are needed. Occurrences of labeled match will be rare, and true beginners are unlikely to encounter them early on.
+
+Labeled match does not introduce arbitrary control flow (like general `goto`) or surprising implicit control flow (like `switch` fallthrough in C and descendants). The mechanism fits nicely into how rust works today.
```

Comment:
I think it’s possible to semantically understand the construct as a reducible CFG similar to `loop { match { … } }`, and not just in the trivial sense that any CFG can technically be made reducible. This would give a reasonable way to apply static analyses that are more naturally expressed in terms of structured control flow (with some loss of precision wrt possible paths through the CFG). If this is not the case, then that’s another drawback in my opinion.

However, it’s true that the sensible MIR lowering for achieving the desired codegen will usually be irreducible from the start. Today you can only write programs that start out as reducible CFG and possibly become irreducible during optimization. Although I don’t know if any current MIR optimizations actually do that (jump threading is the usual suspect) or whether the ball is completely in LLVM’s court for this.

Reply:
If we get the borrowck to understand this then this will really be irreducible

Reply:
On second thought, maybe this isn’t as novel as it seems. With async/await you can also describe what is conceptually irreducible control flow inside a state machine. For example, resuming `loop { g().await; }` after suspending at the `await` jumps into the middle of the loop body (after calling `g().into_future()`, before calling `poll()`) so there isn’t a single header block that dominates the loop. Here, too, it’s possible to pretend the state machine is a big loop + match to make it reducible but this loses precision about which state transitions are possible (e.g., you can’t go “backwards” within one iteration of the loop, but in the loop+match CFG it appears possible unless you do the equivalent of jump threading wrt the state variable). If I’m reading [the lowered MIR](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b118bcd0fece7001327ba3bd8ed029ba) correctly then this already generates irreducible MIR today without optimizations.

Reply:
> If we get the borrowck to understand this then this will really be irreducible

Yes that’s one analysis that would have to choose between precision and dealing with irreducibility. How does it deal with async/await state machines (see my last comment) today?

Reply:
I suspect they're borrow-checked before the coroutine transform but I could be wrong.